### PR TITLE
Provide access to the page meta information

### PR DESCRIPTION
**Pull request by chrisvanrun**
When calling the `.page()` of the APIBase I have a need to access the returned `"count"`.

I could get away with doing a custom page call via URL components to get these values, but then spoofing them in CIRRUS becomes a bit of a hassle and a mess.

The quickest way I could come up with, without changing the return values was to park the page meta data into a referenced dictionary.

### CODE REVIEW

**Comment by chrisvanrun** [comment]
CI is failing because of infrastructure issues. These are addressed in:
- #127

**Review by pkcakeout — COMMENTED** [review]
Oof... I think passing a dict sort-of "by-reference" works, of course, but is a bit of an anti-pattern.

But I would have a proposal for a different implementation, I think is a lot cleaner: Attach the metadata to _everything by default:

~~~~~~~
class PageResult(dict):
    offset: int = 0
    count: int = 0
    # ??? page: int
    
    def __init__(self, offset, count, *args, **kwargs):
        super(PageResult, self).__init__(*args, **kwargs)
        self.offset = offset
        self.count = count

#---------

Then later:

response = yield self.yield_request(
    method="GET", path=self.base_path, params=params
)
results = PageResult(response["offset"], response["count"], response["results"])
~~~~~~~

This way we avoid having to use mutable arguments which are super-messy. We do not change the interface and make the relevant data accessible everywhere. Just provide a positive list of data elements you want to extract. Right now you extract everything except "results" which could also be considered messy because there is no positive data validation happening. We validate everything else through the use of json schema. Would be nice to have the pagination metadata also be validated by providing the list of elements you are actually interested in.

**Comment by chrisvanrun** [comment]
> This way we avoid having to use mutable arguments which are super-messy.
Quick haste generally makes a mess.

I've incorporated your suggestion. I didn't see any neat way to also cover `iterate_all` and ended-up dropping that. Rational is if one decides to just iterate over all items instead of paginating yourself I suspect you don't really care about the total count and you can bet that anyway by retrieving a `.page(limit=1)`.

**Comment by pkcakeout** [comment]
Regarding iterate_all: Yes pretty much what you say. You _could_ create a conditional wrapper to also pass the data along, but I do not expect that this will be every used for the reason you are stating yourself :-)

File: gcapi/apibase.py [hunk 1]
```diff
@@ -176,3 +182,15 @@ def partial_update(self, pk, **kwargs):
 
     def delete(self, pk):
         return (yield from self.perform_request("DELETE", pk=pk))
+
+
+class PageResult(list):
```
Comment (pkcakeout) [inline gcapi/apibase.py:187]: Just out of curiosity. Did you find any references on subclassing from list vs UserList?

**Review by pkcakeout — APPROVED** [review]

File: gcapi/apibase.py (hunk 1 above)
Comment (chrisvanrun) [reply gcapi/apibase.py:187]: It was a bit of a mixed bag. Python2 had a strict suggestion to use UserList, this is not the case for python 3. 

Depending on what behaviour you are looking for the common denotion is now to use a collections (e.g. UserList, or Sequence). I felt it should be a minimum change of the interface so I went with `list`. This did mean that the `count()` couldn't be overwritten, hence the `total_count`.

**Pull request merged**